File: longest_substring_without_repeating_characters.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

int lengthOfLongestSubstring(string s) {
    unordered_map<char, int> char_to_idx;

    int left = -1;
    int right = 0;
    int length = 0;
    while (right != s.length()) {
        if (char_to_idx.find(s[right]) != char_to_idx.end()
            && char_to_idx[s[right]] > left) {
            left = char_to_idx[s[right]];
        }
        char_to_idx[s[right]] = right;
        length = max(length, right - left);
        right++;
    }
    return length;
}

int main() {
    // string s = "bbbbb";
    // std::cout << lengthOfLongestSubstring(s);

    string s = "abcabcbb";
    std::cout << lengthOfLongestSubstring(s);
}

File: python_solutions/longest_substring_without_repeating_characters.py
def lengthOfLongestSubstring(s: str) -> int:
    left = -1
    max_len = 0
    char_to_idx = {}
    for idx, char in enumerate(s):
        if char in char_to_idx and char_to_idx[char] > left:
            left = char_to_idx[char]
        char_to_idx[char] = idx
        max_len = max(max_len, idx - left)
    return max_len


if __name__ == '__main__':
    test_cases = [
        {'input': 'aaaaa', 'output': 1},
        {'input': 'ababab', 'output': 2},
        {'input': 'abcade', 'output': 5},
        {'input': 'pwwkew', 'output': 3},
        {'input': 'abcabcbb', 'output': 3}
    ]

    for test_case in test_cases:
        ans = lengthOfLongestSubstring(test_case['input'])
        print(ans == test_case['output'], ans)
